File: wwwroot/src/raytracer/Ray.ts
import { type Vec3, add, scale } from "./vec3";

export default class Ray {
    #origin: Vec3;
    #direction: Vec3;

    constructor(origin: Vec3, direction: Vec3) {
        this.#origin = origin;
        this.#direction = direction;
    }

    get origin() {
        return this.#origin;
    }

    get direction() {
        return this.#direction;
    }

    at(t: number): Vec3 {
        return add(this.#origin, scale(this.#direction, t));
    }
}

File: rust/src/ray.rs
use crate::vec3::{Point3, Vec3};

pub struct Ray {
    pub origin: Point3,
    pub direction: Vec3,
    pub time: f64
}

impl Ray {
    pub fn at(&self, t: f64) -> Point3 {
        self.origin + t * self.direction
    }
}
